File: exercises/day23-Portal-Stabilization/csharp/PortalStabilization/Models/PortalStabilizer.cs
namespace PortalStabilization.Models;

public class PortalStabilizer
{
    private readonly List<TemporalAnomaly> _processedAnomalies = new();
    
    /// <summary>
    /// Process a batch of temporal anomalies to balance karma and stabilize the portal.
    /// Removes duplicate karma patterns and calculates dimensional karma balance.
    /// </summary>
    public StabilizationReport Stabilize(List<TemporalAnomaly> anomalies)
    {
        var uniqueAnomalies = RemoveDuplicateKarmaPatterns(anomalies);
        var karmaBalances = CalculateKarmaBalances(uniqueAnomalies);
        
        _processedAnomalies.AddRange(uniqueAnomalies);
        
        return new StabilizationReport(
            uniqueAnomalies.Count,
            karmaBalances,
            CalculateKarmaHarmony(karmaBalances)
        );
    }
    
    /// <summary>
    /// Remove duplicate karma patterns from the batch.
    /// An anomaly is a duplicate if it has the same karma signature as another.
    /// </summary>
    private List<TemporalAnomaly> RemoveDuplicateKarmaPatterns(List<TemporalAnomaly> anomalies)
    {
        var unique = new List<TemporalAnomaly>();
        
        foreach (var anomaly in anomalies)
        {
            var isDuplicate = false;
            
            // Check against all already processed anomalies
            foreach (var processed in _processedAnomalies)
            {
                if (anomaly.HasSameKarmaAs(processed))
                {
                    isDuplicate = true;
                    break;
                }
            }
            
            // Check against anomalies in current batch
            if (!isDuplicate)
            {
                foreach (var existing in unique)
                {
                    if (anomaly.HasSameKarmaAs(existing))
                    {
                        isDuplicate = true;
                        break;
                    }
                }
            }
            
            if (!isDuplicate)
            {
                unique.Add(anomaly);
            }
        }
        
        return unique;
    }
    
    /// <summary>
    /// Calculate karma balance for each dimension.
    /// </summary>
    private List<KarmaBalance> CalculateKarmaBalances(List<TemporalAnomaly> anomalies)
    {
        var dimensions = new List<string>();
        
        // Find all unique dimensions
        foreach (var anomaly in anomalies)
        {
            var found = false;
            foreach (var dim in dimensions)
            {
                if (dim == anomaly.Dimension)
                {
                    found = true;
                    break;
                }
            }
            if (!found)
            {
                dimensions.Add(anomaly.Dimension);
            }
        }
        
        // Calculate karma balance for each dimension
        var balances = new List<KarmaBalance>();
        foreach (var dimension in dimensions)
        {
            var totalKarmaDebt = 0;
            var count = 0;
            
            foreach (var anomaly in anomalies)
            {
                if (anomaly.Dimension == dimension)
                {
                    totalKarmaDebt += anomaly.KarmaDebt;
                    count++;
                }
            }
            
            balances.Add(new KarmaBalance(dimension, totalKarmaDebt, count));
        }
        
        return balances;
    }
    
    /// <summary>
    /// Calculate overall karma harmony across all dimensions.
    /// Lower average karma debt = higher harmony = more stable portal.
    /// </summary>
    private double CalculateKarmaHarmony(List<KarmaBalance> balances)
    {
        if (balances.Count == 0)
        {
            return 100.0;
        }
        
        var totalAverageKarmaDebt = 0.0;
        foreach (var balance in balances)
        {
            totalAverageKarmaDebt += balance.AverageKarmaDebt();
        }
        
        var meanKarmaDebt = totalAverageKarmaDebt / balances.Count;
        return Math.Max(0.0, 100.0 - (meanKarmaDebt * 10));
    }
    
    public int GetTotalProcessedCount() => _processedAnomalies.Count;
}

File: exercises/day23-Portal-Stabilization/typescript/src/PortalStabilizer.ts
import { TemporalAnomaly, hasSameKarmaAs } from './models/TemporalAnomaly';
import { KarmaBalance, averageKarmaDebt } from './models/KarmaBalance';
import { StabilizationReport } from './models/StabilizationReport';

export class PortalStabilizer {
  private processedAnomalies: TemporalAnomaly[] = [];

  /**
   * Process a batch of temporal anomalies to balance karma and stabilize the portal.
   * Removes duplicate karma patterns and calculates dimensional karma balance.
   */
  stabilize(anomalies: TemporalAnomaly[]): StabilizationReport {
    const uniqueAnomalies = this.removeDuplicateKarmaPatterns(anomalies);
    const karmaBalances = this.calculateKarmaBalances(uniqueAnomalies);

    this.processedAnomalies.push(...uniqueAnomalies);

    return {
      processedCount: uniqueAnomalies.length,
      karmaBalances,
      karmaHarmony: this.calculateKarmaHarmony(karmaBalances),
    };
  }

  /**
   * Remove duplicate karma patterns from the batch.
   * An anomaly is a duplicate if it has the same karma signature as another.
   */
  private removeDuplicateKarmaPatterns(
    anomalies: TemporalAnomaly[]
  ): TemporalAnomaly[] {
    const unique: TemporalAnomaly[] = [];

    for (const anomaly of anomalies) {
      let isDuplicate = false;

      // Check against all already processed anomalies
      for (const processed of this.processedAnomalies) {
        if (hasSameKarmaAs(anomaly, processed)) {
          isDuplicate = true;
          break;
        }
      }

      // Check against anomalies in current batch
      if (!isDuplicate) {
        for (const existing of unique) {
          if (hasSameKarmaAs(anomaly, existing)) {
            isDuplicate = true;
            break;
          }
        }
      }

      if (!isDuplicate) {
        unique.push(anomaly);
      }
    }

    return unique;
  }

  /**
   * Calculate karma balance for each dimension.
   */
  private calculateKarmaBalances(
    anomalies: TemporalAnomaly[]
  ): KarmaBalance[] {
    const dimensions: string[] = [];

    // Find all unique dimensions
    for (const anomaly of anomalies) {
      let found = false;
      for (const dim of dimensions) {
        if (dim === anomaly.dimension) {
          found = true;
          break;
        }
      }
      if (!found) {
        dimensions.push(anomaly.dimension);
      }
    }

    // Calculate karma balance for each dimension
    const balances: KarmaBalance[] = [];
    for (const dimension of dimensions) {
      let totalKarmaDebt = 0;
      let count = 0;

      for (const anomaly of anomalies) {
        if (anomaly.dimension === dimension) {
          totalKarmaDebt += anomaly.karmaDebt;
          count++;
        }
      }

      balances.push({
        dimension,
        totalKarmaDebt,
        anomalyCount: count,
      });
    }

    return balances;
  }

  /**
   * Calculate overall karma harmony across all dimensions.
   * Lower average karma debt = higher harmony = more stable portal.
   */
  private calculateKarmaHarmony(balances: KarmaBalance[]): number {
    if (balances.length === 0) {
      return 100.0;
    }

    let totalAverageKarmaDebt = 0.0;
    for (const balance of balances) {
      totalAverageKarmaDebt += averageKarmaDebt(balance);
    }

    const meanKarmaDebt = totalAverageKarmaDebt / balances.length;
    return Math.max(0.0, 100.0 - meanKarmaDebt * 10);
  }

  getTotalProcessedCount(): number {
    return this.processedAnomalies.length;
  }
}
